### ISSUE TITLE
Allow Ref Studio to run locally as a web app

### ISSUE THREAD

**Issue by hammer** [refstudio/refstudio#347]
Order of operations here, at a high level, maybe:
- Refactor Python sidecar to run as a service or set of services
- RS Desktop frontend talks to sidecar service(s) over websockets
- RS frontend runs within web server

I'm mostly interested in local deployment for now, we can file a separate issue to host it somewhere once we get local deployment working.

**Comment by hammer** [comment]
Reminder that [#17](https://github.com/refstudio/refstudio/issues/17) had some discussion of sidecar over websockets

**Comment by cguedes** [comment]
This is the current version of the app architecture changes to support desktop and web version of RefStudio. 

![image](https://github.com/refstudio/refstudio/assets/174127/c97553bd-00df-45e8-a37c-e5979e8851bf)


**Sidecar**

- [ ] HTTP server for the python backend commands
     - ex: `POST /ingest/upload { body }`, `GET /ingest/references`, `GET /ingest/status`, `POST chat/...`, `GET /search/...`
- [ ] Update ingestion API to receive the PDF files instead of PDF files path
     - we should support both methods (PDF files uploaded and the current path) for now
     - for the PDF files upload version, we need an additional parameter with the project name (or id) to support multiple projects. That would be useful to scope the multiple `references.json` files in the backend server supporting filesystem
- [ ] Frontend `sidecar.ts` facade that calls Tauri or HTTP backend APIs
     - Maybe we can (should) rename sidecar to something more agnostic
- [ ] Reimplement flow for PDF uploads to work with the new HTTP backend API

**Application Launcher**

- [ ] Run the application inside a web server
    - We can start by running `yarn dev` to run vite server (change initialised to true in the AppStartup to test)
    - We might remove the splash-screen to simplify code and implement a "html/react/native" splash-screen
    - We might use Vite `import....` to conditionally build the app and have code splitting using System.Import
- [ ] Adjust SettingsManager to work in web
    - In web, we need to not read env variables (or we can have a Tauri facade for web that just "noop" env vars)
    - In web, we can't load/save settings to a file

**Filesystem**

- [ ] HTTP server (python) for a virtual filesystem
- [ ] Facade that supports remote filesystem to work in the web
- [ ] Facade that supports local filesystem to work in the desktop Tauri

**Menu & Events**

- [ ] Explore the support for dynamically create (and update) Tauri menu from the TS code
- [ ] Menu component for the web version
- [ ] Facade to support events in the web version


**Project & files lifecycle**

* TODO


Note: we will add more tasks as we know more about how to "decouple" from Tauri internal APIs)

**Comment by danvk** [comment]
Summary of huddle today with @danvk @cguedes and @gjreda (feel free to adjust, this is just from my memory):

- Three things to discuss re: API: PDF ingestion, settings, filesystem
- PDF ingestion
  - Carlos: We should include a project identifier on all requests
  - Greg: Ingest endpoint does some copying of PDFs anyway so the required changes here will be minimal.
  - Dan: Putting binary PDFs in JSON isn't going to work, need to use FormData
- Settings
  - Carlos: We currently use `tauri-settings`. It doesn't do much for us, though; migrating settings was annoying.
  - Dan: We should move all settings logic into the HTTP server. This is a place where TS touches the filesystem but shouldn't.
  - Carlos: We'd prefer not to include things like `OPEN_API_KEY` on HTTP requests
  - Dan: Let's just send that once (when the user configures it) and then the Python API handlers can read it from disk/database, it never needs to go over the wire again.
  - Carlos: is it useful to have a nested settings schema? Probably not.
  - Greg will make a Python equivalent of `settingsManager.ts`
- Filesystem
  - Dan: there are 9 functions in `filesystem.ts`. Easiest path forward is just to implement those in an HTTP API.
  - Carlos: let's get rid of path manipulation as a thing in the frontend (big +1 from Dan)
  - Dan/Carlos: might eventually make sense to switch to a higher-level API but implementing an FS API for now will unblock us.
  - Dan: latency with localhost is really good, so not too concerned about doing lots of sequential async FS calls

Greg will push on merging #359 and then adding these APIs. Dan will get #358 to a mergeable state by end of day and hopefully the APIs will be ready to use by next Wednesday.

**Comment by gjreda** [comment]
@hammer closing this one, but feel free to reopen. Largely fixed by #425
